### DISCUSSION TITLE
Detecting hover events on regions covered by another interactive widget

### DISCUSSION THREAD

**Discussion by ennis** [emilk/egui#3450] [Q&A]
I have a custom list widget, where each row is composed of a background rectangle, which changes color on hover, and a text label on top of it, which can be double-clicked to edit it:

```rust
// Row background
let width = ui.available_width();
let size = vec2(width, 20.0);
let (rect, response) = ui.allocate_exact_size(size, Sense::hover());

// change row background color if hovered
let bg_color = if response.hovered { ... } else { ... }; 

{
  // Text label for the row + double-click to rename
  let text_edit_rect = Rect::from_two_pos(rect.left_top() + vec2(36.0, 0.0), rect.right_bottom() - vec2(16.0, 0.0));  // subregion of the row rect
  let resp = ui.allocate_rect(text_edit_rect, Sense::click_and_drag());
  if resp.double_clicked() {
          state.editing_title = true;
          // ... show text edit when double-clicked ...
  }
}
```

When I do this, the row background doesn't receive "hovered" when I move the mouse over the text label, presumably because the text label covering the row background is considered "interactive". If I use `Sense::hover` for the text label I do get the hover event, but I need to sense clicks on the label, in order to implement "double-click" to rename.

In short, I'm looking for a way to get `hovered` responses for a "parent" widget even when it is covered by an interactive widget. My use case here is to change the background color of some element when the mouse is over it, even if there are interactive widgets on top of it.

Is there a simple/principled way to do that? So far I've considered:
- OR'ing the `hovered()` response of all child elements with the `hovered()` response of the row background. However it's easy to forget to do that when adding a new button on the row.
- Using `response.rect.contains(pointer_pos)` as suggested by the docs, but I'm not sure if that's a good solution: for instance, what will happen when clicking and dragging an unrelated DragValue in the UI, and, while dragging, I move the pointer over the row background? will it be highlighted even though the mouse pointer is "captured" by the DragValue?

**Comment by JumpyLionnn** [comment]
if the inner widgets are not draggable you can try to use this:
```rust
let rect = response.rect;
let layer_id = response.layer_id;
let is_hovering = {
    rect.is_positive() && {
        let pointer_pos = ui.ctx().input(|i| i.pointer.interact_pos());
        if let Some(pointer_pos) = pointer_pos {
            rect.contains(pointer_pos) && ui.ctx().layer_id_at(pointer_pos) == Some(layer_id)
        } else {
            false
        }
    }
};
let not_dragging = ui.memory(|reader| reader.is_anything_being_dragged() && !reader.is_being_dragged(response.id)); // you can check here for inner widget ids
if is_hovering && not_dragging {
    // do whatever you want
}
```
It doesn't account for clicks. The memory doesn't provide this information.
If your inner widgets are draggable then you will need to track each one of them to check what is dragged

I hope this helps :)
